### ISSUE TITLE
Docker starts containers with restart policy `on-failure` after host reset

### ISSUE THREAD

**Issue by sammy007** [moby/moby#47846]
### Description

It's a host failure, not a container failure so `none` policy should apply in such case. There is literally no way to make docker to act as supervisor for a container which only restarts it in case of **container failure**.

### Reproduce

1. create docker container with on-failure policy
2. restart docker
3. docker ps => 0 containers

a. create docker container with on-failure policy
b. reset HOST
c. docker ps => 1 container

### Expected behavior

Do not start container with restart policy set to `on-failure` after host reset.

### docker version

```bash
Client: Docker Engine - Community
 Version:           26.1.3
 API version:       1.45
 Go version:        go1.21.10
 Git commit:        b72abbb
 Built:             Thu May 16 08:33:42 2024
 OS/Arch:           linux/amd64
 Context:           default
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    26.1.3
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.14.0
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.27.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
```

**Comment by ledlamp** [comment]
You mean if you restart docker daemon it does not restart container but if you reboot the host it does restart container?

**Comment by ledlamp** [comment]
I reproduced this

![Image](https://github.com/user-attachments/assets/d9da7826-c4c1-49f8-ab7a-deba5600890f)

When either restarting docker daemon or restarting host, the container with --restart=on-failure is restarted.

But the documentation says "It doesn't restart the container if the daemon restarts."

![Image](https://github.com/user-attachments/assets/a5defcef-7f55-4b40-b09d-7c36b1769d41)

The documentation is wrong.

Docker version:

```
Client:
 Version:           26.1.3
 API version:       1.45
 Go version:        go1.22.2
 Git commit:        26.1.3-0ubuntu1~24.04.1
 Built:             Mon Oct 14 14:29:26 2024
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          26.1.3
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.22.2
  Git commit:       26.1.3-0ubuntu1~24.04.1
  Built:            Mon Oct 14 14:29:26 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.24
  GitCommit:        
 runc:
  Version:          1.1.12-0ubuntu3.1
  GitCommit:        
 docker-init:
  Version:          0.19.0
  GitCommit:
```

**Comment by mindThomas** [comment]
I can also confirm this happening to us. Any work being done on resolving this?

**Comment by tonygoldman** [comment]
I noticed the same behavior when I cut power to my machine.
After a power-loss, all dockers are started automatically.
According to the docs, this is not the intended behavior of on-failure.

**Comment by TheOverpassArsonist** [comment]
I don't know if this is even the right place to comment about this (not fully familiar with what 'moby' is but based on reading the readme I think it might be?) but I am running into this right now and it is genuinely infuriating.

I'm using a simple docker Jellyfin container pointed to a library on a Mdadm RAID array, and every single time I boot, without fail, the container starts before the drive is opened, automatically creates a new folder, as root, to occupy every single bind, and thus blocks the mdadm container from being auto-mounted because the folder it _should_ have automounted to is now a series of directories - directories that it doesn't even have the permission to write to even if it wanted to.

I have no idea why this is happening and I'm 90% sure I tested it before and this _didn't_ happen, but I just rebooted and sure enough it did. "on-failure" acts in the same way as "always" for me as of less than 5 minutes ago. (with a fully uptodate system using the latest from arch repos, because I rebooted after I got home and updated my system.)

**Comment by akerouanton** [comment]
> I don't know if this is even the right place to comment about this (not fully familiar with what 'moby' is but based on reading the readme I think it might be?)

Yes this is the right place. Docker Engine (or sometimes called 'Docker CE') is currently a vanilla build of Moby.

One hypothesis (but I didn't check): when you reboot your server, systemd shuts down the Engine which tries to gracefully stop all running containers.

If a container doesn't correctly handle SIGTERM (or whichever stop signal you set on that container), the Engine waits for a few seconds (10s by default) and then SIGKILL it. If your container exits with a non-zero exit code, either on SIGTERM or due to the SIGKILL sent by the Engine, on next boot the Engine will consider that container failed and will restart it.

You may need to review how your containers are handling SIGTERM, and which exit code they return.

> I'm using a simple docker Jellyfin container pointed to a library on a Mdadm RAID array, and every single time I boot, without fail, the container starts before the drive is opened,

Independently of signal and exit codes, you should probably tweak `docker.service` unit file to ensure it has a `After=` clause (or equivalent) that prevents the Engine from starting before your mountpoints are created.
